File: .config/ags/js/left.js
const hyprland = await Service.import("hyprland");
const mpris = await Service.import("mpris");

function Workspaces() {
  const activeId = hyprland.active.workspace.bind("id");
  const workspaces = hyprland.bind("workspaces").as((ws) =>
    ws
      .sort((a, b) => a.id - b.id)
      .map(({ id }) =>
        Widget.Label({
          className: activeId.as((i) => `${i === id ? "active" : ""}`),
          label: activeId.as((i) => `${i === id ? "X" : "O"}`),
        }),
      ),
  );

  return Widget.Box({
    className: "workspaces",
    children: workspaces,
    spacing: 10,
  });
}

function ClientTitle() {
  return Widget.Label({
    className: "client",
    maxWidthChars: 25,
    ellipsize: true,
    truncate: "end",
    label: hyprland.active.client.bind("title").as((name) => name || "(-_-)"),
  });
}

function Media() {
  const trackTitle = Utils.watch(
    null,
    mpris,
    "player-changed",
    () => {
      let active = mpris.players.find(p => p.play_back_status === "Playing") ||
        mpris.players.find(p => p.play_back_status === "Paused")

      return active.track_title
    }
  )

  const artist = Utils.watch(
    null,
    mpris,
    "player-changed",
    () => {
      let active = mpris.players.find(p => p.play_back_status === "Playing") ||
        mpris.players.find(p => p.play_back_status === "Paused")

      return "by " + active.track_artists?.join(", ")
    }
  )
  const icon = Utils.watch(
    null,
    mpris,
    "player-changed",
    () => {
      let active = mpris.players.find(p => p.play_back_status === "Playing") ||
        mpris.players.find(p => p.play_back_status === "Paused")

      return active?.play_back_status == "Playing" ? "" : ""
    }
  )

  const revealArtist = Utils.watch(
    null,
    mpris,
    "player-changed",
    () => {
      let active = mpris.players.find(p => p.play_back_status === "Playing") ||
        mpris.players.find(p => p.play_back_status === "Paused")

      return !!active.track_artists?.length &&
        active.track_artists?.join("").length > 1
    }
  )
  const className = mpris
    .bind("players")
    .as((p) => "media " + (p.length === 0 ? "media_hidden" : ""));

  const visible = mpris.bind("players").as((p) => p.length > 0);

  let Title = function (invisible) {
    return Widget.Label({
      className: "title",
      css: invisible && "color: transparent; text-shadow: none;" || "",
      label: trackTitle,
      visible,
      maxWidthChars: 20,
      ellipsize: true,
      truncate: "end",
    })
  }

  return Widget.Box({
    className,
    spacing: 0,
    visible,
    children: [
      Widget.Box(),
      Widget.Revealer({
        revealChild: visible,
        transition: "crossfade",
        child: Widget.Box({
          spacing: 10,
          children: [
            Widget.Label({ className: "icon", visible, label: icon, }),
            // using overlay here for that widget to not extend the bar height
            // .. happens sometimes when there are jap. / chinese symbols in text
            Widget.Overlay({
              // invisible duplicate of text, keeps widget width
              child: Title(true),
              overlays: [
                Title(),
                Widget.Revealer({
                  transition: "slide_down",
                  transitionDuration: 250,
                  revealChild: revealArtist,
                  child: Widget.Label({
                    className: "artist",
                    label: artist,
                    hpack: "start",
                    visible,
                  }),
                }),
              ],
            }),
          ],
        }),
      }),
    ],
  });
}

// layout of the bar
export default function Left() {
  return Widget.Box({
    spacing: 8,
    children: [Workspaces(), ClientTitle(), Media()],
  });
}

File: .config/astal/widget/Left.lua
local astal = require("astal")
local Widget = require("astal.gtk3.widget")
local Mpris = require("lgi").require("AstalMpris")
local Hyprland = require("lgi").require("AstalHyprland")
local Pango = require("lgi").require("Pango")
-- local Variable = astal.Variable

local bind = astal.bind
local map = require("lib").map

local hypr = Hyprland.get_default()

local function Workspaces()
  return Widget.Box({
    class_name = "workspaces",
    spacing = 10,
    bind(hypr, "workspaces"):as(function(wss)
      table.sort(wss, function(a, b)
        return a.id < b.id
      end)

      return map(wss, function(ws)
        return Widget.Label({
          class_name = bind(hypr, "focused-workspace"):as(function(fw)
            return fw == ws and "active" or ""
          end),
          label = bind(hypr, "focused-workspace"):as(function(fw)
            return fw == ws and "X" or "O"
          end),
        })
      end)
    end),
  })
end

local function ClientTitle()
  local focused = bind(hypr, "focused-client")

  return focused:as(function(client)
    return Widget.Label({
      class_name = "client",
      max_width_chars = 25,
      ellipsize = Pango.EllipsizeMode.END,
      -- truncate = true, -- TODO: why its undefined?
      label = bind(client, "title"):as(tostring),
    })
  end)
end

-- local function Media()
-- 	return Widget.Box({
--       class_name: ""
--   })
-- end

-- local function Media()
--   local mpris = Mpris.get_default()
--   local players = bind(mpris, "players")
--   -- print(#players)

--   local player = Variable(players[1])
--   local visible = bind(mpris, "players"):as(function(ps) return #ps > 0 end)

-- for _, p in ipairs(players) do
--   -- 0 - playing; 1 - paused; 2 - stopped (https://aylur.github.io/libastal/mpris/enum.PlaybackStatus.html)
--   print(p.playback_status)
--   if p.playback_status == 0 then
--     player = Variable(p)
--     break
--   end
-- end

-- print(player.title)

-- if not active then
--     for _, player in ipairs(players) do
--         if player.play_back_status == "Paused" then
--             active = player
--             break
--         end
--     end
-- end

-- return Widget.Box()
-- return Widget.Box({
--   class_name = #players == 0 and "media_hidden" or "",
--   spacing = 0,
--   visible,
--   valign = "CENTER",
--   Widget.Box(),
--   Widget.Revealer({
--     reveal_child = visible,
--     -- transition = "crossface",
--     Widget.Box({
--       spacing = 10,
--       Widget.Label({
--         class_name = "icon",
--         visible,
--         label = bind(player, "playback_status"):as(function(s)
--           return s == 0 and "" or ""
--         end),
--       }),
--       Widget.Label({
--         class_name = "title",
--         visible,
--         max_width_chars = 20,
--         -- truncate = true,
--         label = bind(player, "title"):as(tostring)
--       })
--     })
--   }),
-- })
-- end

local function Media()
  local mpris = Mpris.get_default()
  return Widget.Box({
    bind(mpris, "players"):as(function(players)
      local player = players[1]
      if player then
        return Widget.Box({
          Widget.Label({ label = bind(player, "title") }),
        })
      end
    end),
  })
end

return function()
  return Widget.Box({
    spacing = 8,
    Workspaces(),
    ClientTitle(),
    -- Media(),
  })
end
